lib/experimental/interactivity-api/directive-processing.php

```diff
@@ -28,37 +28,258 @@ function gutenberg_interactivity_mark_root_blocks( $parsed_block, $source_block,
+ *
+ * @return string The processed HTML.
+ */
+function gutenberg_process_non_interactive_block( $non_interactive_block, $context, $directives ) {
+	$block_index = 0;
+	$content     = '';
+	foreach ( $non_interactive_block['innerContent'] as $inner_content ) {
+		if ( is_string( $inner_content ) ) {
+			// This content belongs to a non interactive block and therefore it cannot
+			// contain directives. We add the HTML directly to the final output.
+			$content .= $inner_content;
+		} else {
+			// This is an inner block. It may be an interactive block or a
+			// non-interactive block.
+			$inner_block = $non_interactive_block['innerBlocks'][ $block_index++ ];
+
+			if ( 'core/interactivity-wrapper' === $inner_block['blockName'] ) {
+				$content .= gutenberg_process_interactive_block( $inner_block, $context, $directives );
+			} elseif ( 'core/non-interactivity-wrapper' === $inner_block['blockName'] ) {
+				$content .= gutenberg_process_non_interactive_block( $inner_block, $context, $directives );
+			}
+		}
+	}
+	return $content;
+}
+
+/**
+ * Processes interactive HTML by applying directives to the HTML tags.
+ *
+ * It uses the WP_Directive_Processor class to parse the HTML and apply the
+ * directives. The directives are specified in the $directives array and are
+ * applied to the tags that have the corresponding data attribute. If a tag
+ * contains a 'WP-INNER-BLOCKS' string and there are inner blocks to process,
+ * the function processes these inner blocks and replaces the 'WP-INNER-BLOCKS'
+ * tag in the HTML with those blocks.
+ *
+ * @param string $html The HTML to process.
+ * @param mixed  $context The context to use when processing.
+ * @param array  $directives The directives to apply.
+ * @param array  $inner_blocks The inner blocks to process.
+ *
+ * @return string The processed HTML.
+ */
+function gutenberg_process_interactive_html( $html, $context, $directives, $inner_blocks = array() ) {
+	$tags                   = new WP_Directive_Processor( $html );
+	$prefix                 = 'data-wp-';
+	$tag_stack              = array();
+	$inner_processed_blocks = array();
+	$inner_blocks_index     = 0;
+	while ( $tags->next_tag( array( 'tag_closers' => 'visit' ) ) ) {
+		$tag_name = $tags->get_tag();
+
+		// Processes the inner blocks.
+		if ( str_contains( $tag_name, 'WP-INNER-BLOCKS' ) && ! empty( $inner_blocks ) && ! $tags->is_tag_closer() ) {
+			if ( 'core/interactivity-wrapper' === $inner_blocks[ $inner_blocks_index ]['blockName'] ) {
+				$inner_processed_blocks[ strtolower( $tag_name ) ] = gutenberg_process_interactive_block( $inner_blocks[ $inner_blocks_index++ ], $context, $directives );
+			} elseif ( 'core/non-interactivity-wrapper' === $inner_blocks[ $inner_blocks_index ]['blockName'] ) {
+				$inner_processed_blocks[ strtolower( $tag_name ) ] = gutenberg_process_non_interactive_block( $inner_blocks[ $inner_blocks_index++ ], $context, $directives );
+			}
+		}
+		if ( $tags->is_tag_closer() ) {
```

Comment:
This cannot be an `elseif`, otherwise the next else is conditioned by the first condition `! empty( $inner_blocks`. And we need to parse a tag closer of a block without inner_blocks.

Test `test_directives_are_processed_at_tag_end` contemplates this situation.

cc: @luisherranz

Reply:
Ohhhh, ok. That's because we need to add the opener tag to the stack, otherwise, the tag closer removes an extra item from the stack, and we unbalance the tags, right?

I guess in the future we could skip the processing of both the `wp-inner-block` opener and closer tags, but let's leave it with the `if` for now. Thanks, Carlos! 🙂